[CIR] Upstream binary assignments and comma

mmha opened the pull request:
This patch adds `VisitBinAssign` and `VisitBinComma` to the ClangIR `ScalarExprEmitter` to enable assignments and the comma operator.

mmha commented:
cc @erichkeane @andykaylor @dkolsen-pgi

erichkeane reviewed (commented):
A little more testing on comma would be nice, else seems reasonable.

mmha committed:
Add one more comma test case

mmha commented on clang/test/CIR/CodeGen/comma.c:
```diff
@@ -0,0 +1,53 @@
+    float f;
+    int i;
+
+    b = true, c = 65, f = 3.14f, i = 42;
```
So, I did add another test for `i = 100, 200;` but we don't support ParenExprs, yet, so I can't really check that the result of that expression is `200` for now.

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp:
```diff
@@ -807,6 +808,65 @@ class ScalarExprEmitter : public StmtVisitor<ScalarExprEmitter, mlir::Value> {
+
+  mlir::Value VisitBinComma(const BinaryOperator *e) {
+    cgf.emitIgnoredExpr(e->getLHS());
+    // NOTE: We don't need to EnsureInsertPoint() like LLVM codegen.
```
I think this is a difference in the way LLVM IR and MLIR are generated. You can create LLVM IR instructions in the void and insert them later. MLIR operations are always created in some region or block. The classic codegen calls `EnsureInsertPoint` to create a basic block to receive emitted instructions. It calls it all over the place. I'm not sure why we'd have a comment about it here. This is the only reference to `EnsureInsertPoint` in the incubator code.

andykaylor commented on clang/lib/CIR/CodeGen/CIRGenFunction.h:
```diff
@@ -572,6 +575,11 @@ class CIRGenFunction : public CIRGenTypeCache {
 
+  /// Given an assignment `*lhs = rhs`, emit a test that checks if \p rhs is
+  /// nonnull, if 1\p LHS is marked _Nonnull.
+  void emitNullabilityCheck(LValue lhs, mlir::Value rhs,
```
This is out of order alphabetically. A few other things are, so we need to do a cleanup pass anyway.

andykaylor reviewed (approved):
This looks good to me. I have a few comments, but nothing that needs to be changed immediately.

mmha commented on clang/lib/CIR/CodeGen/CIRGenExprScalar.cpp:
```diff
@@ -807,6 +808,65 @@ class ScalarExprEmitter : public StmtVisitor<ScalarExprEmitter, mlir::Value> {
+    mlir::Value rhs;
+    LValue lhs;
+
+    switch (e->getLHS()->getType().getObjCLifetime()) {
```
Hmm, I don't know about the performance implications (I'd expect them to be insignificant) but I find your suggestion to be less readable. I still need a `case Qualifiers::OCL_None:` with an `llvm_unreachable()` to satisfy clang-tidy and it just adds more branches to the function to follow

mmha commented on clang/lib/CIR/CodeGen/CIRGenExpr.cpp:
```diff
@@ -219,6 +219,13 @@ void CIRGenFunction::emitStoreOfScalar(mlir::Value value, Address addr,
 }
 
+void CIRGenFunction::emitStoreThroughBitfieldLValue(RValue src, LValue dst,
+                                                    mlir::Value &result) {
```
I don't and you raise a good point. OGCG uses a `llvm::Value **` here as an optional return value and it makes sense there because that results in fewer dead instruction emitted if the result is discarded but for CIR that doesn't apply.

I changed the signature.

mmha committed:
Change signature of emitStoreThroughBitfieldLValue

mmha committed:
[CIR] Upstream binary assignments and comma

This patch adds `VisitBinAssign` and `VisitBinComma` to the ClangIR `ScalarExprEmitter` to enable assignments and the comma operator.

mmha committed:
clang-format

mmha committed:
Merge branch 'main' into cir-binassign
